File: tests/client-gen/example_program_gen/types/foo_struct.py
from __future__ import annotations
from . import bar_struct, foo_enum
import typing
from construct import Container
import borsh_construct as borsh


class FooStructFields(typing.TypedDict):
    field1: int
    field2: int
    nested: bar_struct.BarStructFields
    vec_nested: list[bar_struct.BarStructFields]
    option_nested: typing.Optional[bar_struct.BarStructFields]
    enum_field: foo_enum.FooEnumKind


class FooStructJSON(typing.TypedDict):
    field1: int
    field2: int
    nested: bar_struct.BarStructJSON
    vec_nested: list[bar_struct.BarStructJSON]
    option_nested: typing.Optional[bar_struct.BarStructJSON]
    enum_field: foo_enum.FooEnumJSON


class FooStruct:
    layout = borsh.CStruct(
        "field1" / borsh.U8,
        "field2" / borsh.U16,
        "nested" / bar_struct.BarStruct.layout,
        "vec_nested" / borsh.Vec(bar_struct.BarStruct.layout),
        "option_nested" / borsh.Option(bar_struct.BarStruct.layout),
        "enum_field" / foo_enum.layout,
    )

    def __init__(self, fields: FooStructFields) -> None:
        self.field1 = fields["field1"]
        self.field2 = fields["field2"]
        self.nested = bar_struct.BarStruct(**fields["nested"])
        self.vec_nested = list(
            map(lambda item: bar_struct.BarStruct(**item), fields["vec_nested"])
        )
        self.option_nested = (
            fields["option_nested"] and bar_struct.BarStruct(**fields["option_nested"])
        ) or None
        self.enum_field = fields["enum_field"]

    @classmethod
    def from_decoded(cls, obj: Container) -> "FooStruct":
        return cls(
            FooStructFields(
                field1=obj.field1,
                field2=obj.field2,
                nested=bar_struct.BarStruct.from_decoded(obj.nested),
                vec_nested=list(
                    map(
                        lambda item: bar_struct.BarStruct.from_decoded(item),
                        obj.vec_nested,
                    )
                ),
                option_nested=(
                    obj.option_nested
                    and bar_struct.BarStruct.from_decoded(obj.option_nested)
                )
                or None,
                enum_field=foo_enum.from_decoded(obj.enum_field),
            )
        )

    @classmethod
    def to_encodable(cls, fields: FooStructFields) -> dict[str, typing.Any]:
        return {
            "field1": fields["field1"],
            "field2": fields["field2"],
            "nested": bar_struct.BarStruct.to_encodable(fields["nested"]),
            "vec_nested": list(
                map(
                    lambda item: bar_struct.BarStruct.to_encodable(item),
                    fields["vec_nested"],
                )
            ),
            "option_nested": (
                fields["option_nested"]
                and bar_struct.BarStruct.to_encodable(fields["option_nested"])
            )
            or None,
            "enum_field": fields["enum_field"].to_encodable(),
        }

    def to_json(self) -> FooStructJSON:
        return {
            "field1": self.field1,
            "field2": self.field2,
            "nested": self.nested.to_json(),
            "vec_nested": list(map(lambda item: item.to_json(), self.vec_nested)),
            "option_nested": (self.option_nested and self.option_nested.to_json())
            or None,
            "enum_field": self.enum_field.to_json(),
        }

    @classmethod
    def from_json(cls, obj: FooStructJSON) -> "FooStruct":
        return cls(
            FooStructJSON(
                field1=obj["field1"],
                field2=obj["field2"],
                nested=bar_struct.BarStruct.from_json(obj["nested"]),
                vec_nested=list(
                    map(
                        lambda item: bar_struct.BarStruct.from_json(item),
                        obj["vec_nested"],
                    )
                ),
                option_nested=(
                    obj["option_nested"]
                    and bar_struct.BarStruct.from_json(obj["option_nested"])
                )
                or None,
                enum_field=foo_enum.from_json(obj["enum_field"]),
            )
        )

File: ts-reference/tests/example-program-gen/exp/types/FooStruct.ts
import { PublicKey } from "@solana/web3.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh"

export interface FooStructFields {
  field1: number
  field2: number
  nested: types.BarStructFields
  vecNested: Array<types.BarStructFields>
  optionNested: types.BarStructFields | null
  enumField: types.FooEnumKind
}

export interface FooStructJSON {
  field1: number
  field2: number
  nested: types.BarStructJSON
  vecNested: Array<types.BarStructJSON>
  optionNested: types.BarStructJSON | null
  enumField: types.FooEnumJSON
}

export class FooStruct {
  readonly field1: number
  readonly field2: number
  readonly nested: types.BarStruct
  readonly vecNested: Array<types.BarStruct>
  readonly optionNested: types.BarStruct | null
  readonly enumField: types.FooEnumKind

  constructor(fields: FooStructFields) {
    this.field1 = fields.field1
    this.field2 = fields.field2
    this.nested = new types.BarStruct({ ...fields.nested })
    this.vecNested = fields.vecNested.map(
      (item) => new types.BarStruct({ ...item })
    )
    this.optionNested =
      (fields.optionNested &&
        new types.BarStruct({ ...fields.optionNested })) ||
      null
    this.enumField = fields.enumField
  }

  static layout(property?: string) {
    return borsh.struct(
      [
        borsh.u8("field1"),
        borsh.u16("field2"),
        types.BarStruct.layout("nested"),
        borsh.vec(types.BarStruct.layout(), "vecNested"),
        borsh.option(types.BarStruct.layout(), "optionNested"),
        types.FooEnum.layout("enumField"),
      ],
      property
    )
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  static fromDecoded(obj: any) {
    return new FooStruct({
      field1: obj.field1,
      field2: obj.field2,
      nested: types.BarStruct.fromDecoded(obj.nested),
      vecNested: obj.vecNested.map((item) => types.BarStruct.fromDecoded(item)),
      optionNested:
        (obj.optionNested && types.BarStruct.fromDecoded(obj.optionNested)) ||
        null,
      enumField: types.FooEnum.fromDecoded(obj.enumField),
    })
  }

  static toEncodable(fields: FooStructFields) {
    return {
      field1: fields.field1,
      field2: fields.field2,
      nested: types.BarStruct.toEncodable(fields.nested),
      vecNested: fields.vecNested.map((item) =>
        types.BarStruct.toEncodable(item)
      ),
      optionNested:
        (fields.optionNested &&
          types.BarStruct.toEncodable(fields.optionNested)) ||
        null,
      enumField: fields.enumField.toEncodable(),
    }
  }

  toJSON(): FooStructJSON {
    return {
      field1: this.field1,
      field2: this.field2,
      nested: this.nested.toJSON(),
      vecNested: this.vecNested.map((item) => item.toJSON()),
      optionNested: (this.optionNested && this.optionNested.toJSON()) || null,
      enumField: this.enumField.toJSON(),
    }
  }

  static fromJSON(obj: FooStructJSON): FooStruct {
    return new FooStruct({
      field1: obj.field1,
      field2: obj.field2,
      nested: types.BarStruct.fromJSON(obj.nested),
      vecNested: obj.vecNested.map((item) => types.BarStruct.fromJSON(item)),
      optionNested:
        (obj.optionNested && types.BarStruct.fromJSON(obj.optionNested)) ||
        null,
      enumField: types.FooEnum.fromJSON(obj.enumField),
    })
  }

  toEncodable() {
    return FooStruct.toEncodable(this)
  }
}
